prereise/gather/demanddata/transportation_electrification/generate_scaling_factors.py

```diff
@@ -0,0 +1,211 @@
+    )
+
+    return (
+        df_ua.squeeze().loc[lambda x: x != "[no data]"],
+        df_state.rename(index=state2abv).squeeze(),
+    )
+
+
+def calculate_vmt_for_ua(census_ua, tht_ua):
+    """Calculate the total annual Vehicle Miles Traveled (VMT) in urban areas
+
+    :param dict census_ua: dictionary as returned by :func:`load_census_ua`
+    :param pandas.Series tht_ua: vmt per capita in urban areas as returned by
+        :func:`load_dot_vmt_per_capita`
+    :return: (*dict*) -- keys are state abbreviations and values are series giving
+        annual vmt by urban areas.
+    """
+
+    tht_ua_format = tht_ua.copy()
+    vmt_for_ua = {}
+
+    tht_ua_format.index = tht_ua_format.index.str.replace(r"[, -.]", "", regex=True)
+    format2original = dict(zip(tht_ua_format.index, tht_ua.index))
+
+    for s in census_ua:
+        census_ua_format = census_ua[s].copy()
+        census_ua_format.index = census_ua_format.index.str.replace(
+            r"[, -.]", "", regex=True
+        )
+        common = set(tht_ua_format.index).intersection(set(census_ua_format.index))
+        vmt_for_ua[s] = (
+            pd.DataFrame(
+                {
+                    "Annual VMT": [
+                        365 * tht_ua_format.loc[i] * census_ua_format.loc[i]
+                        for i in common
+                    ]
+                },
+                index=list(common),
+            )
+            .rename(index=format2original)
+            .squeeze()
+        )
+
+    return vmt_for_ua
+
+
+def calculate_vmt_for_state(census_state, tht_state):
+    """Calculate the total annual Vehicle Miles Traveled (VMT) in states
+
+    :param dict census_state: dictionary as returned by :func:`load_census_state`
+    :param pandas.Series tht_state: vmt per capita in states as returned by
+        :func:`load_dot_vmt_per_capita`
+    :return: (*pandas.Series*) -- indices are state abbreviations and values are annual
+        VMT in state.
+    """
+    common = list(set(tht_state.index).intersection(set(census_state.index)))
+    vmt_for_state = tht_state.loc[common] * census_state.loc[common]
+
+    return vmt_for_state
+
+
+def calculate_urban_rural_fraction(vmt_for_ua, vmt_for_state):
+    """Calculate the percentage of Vehicle Miles Traveled (VMT) in urban and rural areas
+
+    :param dict vmt_for_ua: dictionary as returned by :func:`calculate_vmt_for_ua`.
+    :param pandas.Series vmt_for_state: series as returned by
+        :func:`calculate_vmt_for_state`
+    :return: (*tuple*) -- keys are state abbreviations and values are either series of
+        percentage vmt in urban areas (first element) or percentage in rural area
+        (second element)
+    """
+    vmt_for_ua_perc = vmt_for_ua.copy()
+    vmt_for_ra_perc = {}
+    for s in vmt_for_ua:
+        if s in vmt_for_state.index:
+            vmt_for_ua_perc[s] = vmt_for_ua[s] / vmt_for_state.loc[s]
+            vmt_for_ra_perc[s] = 1 - vmt_for_ua_perc[s].sum()
```

Comment:
Might check that these values are between 0 and 1 in case there is an issue with data?

Reply:
All good:
```
>>> for s, v in vmt_ua_perc.items():
...     if s != "DC":
...             print(f"{s}: {v.max()}")
...     else:
...             print(f"{s}: {v}")
... 
AL: 0.14434760346942355
AK: 0.26437207277844105
AZ: 0.44991303311393605
AR: 0.14908502626966288
CA: 0.2982427206720955
CO: 0.36397376132480563
CT: 0.26032021034444786
DE: 0.3423880212011323
DC: 1
FL: 0.22618050635128556
GA: 0.4205194715339222
HI: 0.4555934566561944
ID: 0.15263531331780308
IL: 0.5686896022596508
IN: 0.21845602247508317
IA: 0.11619400114840563
KS: 0.23042077253226498
KY: 0.06590201248930878
LA: 0.1167878455075878
ME: 0.11716040474540913
MD: 0.33409851003528707
MA: 0.5934979073891866
MI: 0.3462470055304457
MN: 0.4069880542887952
MS: 0.14115482249519
MO: 0.2858434202091423
MT: 0.06639716734171276
NE: 0.22775869035927454
NV: 0.48426477080479685
NH: 0.13401929285747738
NJ: 0.47782865005090186
NM: 0.21564355353847547
NY: 0.5389666383119348
NC: 0.10540342948979183
ND: 0.08904652146843647
OH: 0.1287217813988377
OK: 0.2354443740440485
OR: 0.29558155277403003
PA: 0.25133818104564754
RI: 0.9151826365467751
SC: 0.09462195359436316
SD: 0.11024035295479549
TN: 0.1736666305348097
TX: 0.179917688878257
UT: 0.28819452292365755
VT: 0.16083495321895766
VA: 0.20818156599974846
WA: 0.4318657588567066
WV: 0.12309620751263774
WI: 0.18486930899464962
WY: 0.0705196807271032
```

Reply:
Looks good